### PR TITLE
Improve testing, fix typing problems

### PR DESCRIPTION
**Pull request by soswow**

### CODE REVIEW

File: src/video.tsx [hunk 1]
```diff
@@ -71,7 +73,7 @@ const isSafari = typeof navigator !== 'undefined' ? /^((?!chrome|android).)*safa
 
 export class Video extends Component<VideoProps, VideoComponentState> {
   previousVolume: number = 1;
-  mediaRef: RefObject<SourceElement> = React.createRef();
+  mediaRef: RefObject<HTMLVideoElement & HTMLAudioElement> = React.createRef<HTMLVideoElement & HTMLAudioElement>();
```
Comment (soswow) [inline src/video.tsx:76]: TS not liking otherwise. Need to think of a better one

File: package.json [hunk 2]
```diff
@@ -15,10 +15,10 @@
     "@atlaskit/single-select": "^4.0.3",
     "@atlaskit/spinner": "^8.0.0",
     "react-gh-corner": "^1.1.2",
-    "ts-react-toolbox": "^0.1.23"
+    "ts-react-toolbox": "soswow/ts-react-toolbox#e043fec6732a6f8a8d7f3e53197fd51ee0b952b3"
```
Comment (soswow) [inline package.json:18]: Placeholder while https://github.com/zzarcon/ts-react-toolbox/pull/22 is reviewed

File: __tests__/index.tsx [hunk 3]
```diff
@@ -1,35 +1,42 @@
...
-    })) as jest.Mock<any>;
-    const component = mount((
+    }) as jest.Mock<ReturnType<RenderCallback>, Parameters<RenderCallback>>;
+
+    const component = mount<Video, VideoProps, VideoComponentState>((
       <Video src={src} {...props}>
         {children}
       </Video>
     ));
+    const videoActions = children.mock.calls[0][2];
```
Comment (zzarcon) [inline __tests__/index.tsx:22]: I think this is `any`?

File: __tests__/index.tsx [hunk 4]
```diff
@@ -356,74 +363,132 @@ describe('VideoRenderer', () => {
   });
 
   describe('actions', () => {
-    let videoActions: VideoActions;
-    let videoElement: ReactNode;
-    const children: RenderCallback = (element: ReactNode, state: VideoState, actions: VideoActions) => {
```
Comment (zzarcon) [inline __tests__/index.tsx:361]: Thanks ❤️

File: src/video.tsx [hunk 5]
```diff
@@ -31,7 +32,8 @@ export interface VideoActions {
   toggleMute: () => void;
 }
 
-export type RenderCallback = (videoElement: ReactNode, state: VideoState, actions: VideoActions, ref: RefObject<SourceElement>) => ReactNode;
+export type RenderCallback = (videoElement: ReactElement<SourceElement>, state: VideoState, actions: VideoActions, ref: RefObject<SourceElement>) => ReactNode;
```
Comment (zzarcon) [inline src/video.tsx:42]: Is `ReactElement<SourceElement>` backwards compatible with `ReactNode`?

File: __tests__/index.tsx (hunk 3 above)
Comment (soswow) [reply __tests__/index.tsx:22]: not any more.

File: src/video.tsx (hunk 5 above)
Comment (soswow) [reply src/video.tsx:42]: ReactElement is more concreate version of ReactNode. So, yes, I don't think this is a breaking change.

File: src/video.tsx [hunk 6]
```diff
@@ -264,31 +281,48 @@ export class Video extends Component<VideoProps, VideoComponentState> {
...
+      onWaiting: this.onWaiting,
+      crossOrigin: crossOrigin,
+    }
+
+    if (sourceType === 'video') {
+      return children(
+        <video
+          ref={this.videoRef}
+          poster={poster}
+          {...props}
```
Comment (zzarcon) [inline src/video.tsx:308]: Gooood

File: src/video.tsx [hunk 7]
```diff
@@ -264,31 +281,48 @@ export class Video extends Component<VideoProps, VideoComponentState> {
...
-        crossOrigin={crossOrigin}
-      />,
-      videoState,
-      actions,
-      this.mediaRef  
-    );
+    const { videoState, actions } = this;
+    const { sourceType = 'video', poster, src, children, autoPlay, controls, preload, crossOrigin } = this.props;
+
+    const props: Partial<MediaHTMLAttributes<HTMLVideoElement & HTMLAudioElement>> = {
```
Comment (zzarcon) [inline src/video.tsx:287]: hell of a type but nice

File: src/video.tsx [hunk 8]
```diff
@@ -264,31 +281,48 @@ export class Video extends Component<VideoProps, VideoComponentState> {
...
+      autoPlay: autoPlay,
+      onPlay: this.onPlay,
+      onPause: this.onPause,
+      onVolumeChange: this.onVolumeChange,
+      onTimeUpdate: this.onTimeUpdate,
+      onCanPlay: this.onCanPlay,
+      onDurationChange: this.onDurationChange,
+      onError: this.onError,
+      onWaiting: this.onWaiting,
+      crossOrigin: crossOrigin,
```
Comment (zzarcon) [inline src/video.tsx:300]: just `crossOrigin`

File: src/video.tsx [hunk 9]
```diff
@@ -264,31 +281,48 @@ export class Video extends Component<VideoProps, VideoComponentState> {
...
-    );
+    const { videoState, actions } = this;
+    const { sourceType = 'video', poster, src, children, autoPlay, controls, preload, crossOrigin } = this.props;
+
+    const props: Partial<MediaHTMLAttributes<HTMLVideoElement & HTMLAudioElement>> = {
+      src: src,
+      preload: preload,
+      controls: controls,
+      autoPlay: autoPlay,
+      onPlay: this.onPlay,
```
Comment (zzarcon) [inline src/video.tsx:292]:
```
src,
preload,
controls,
autoPlay,
```

File: src/video.tsx [hunk 10]
```diff
@@ -173,36 +182,46 @@ export class Video extends Component<VideoProps, VideoComponentState> {
...
+    this.currentElement && (this.currentElement.volume = volume);
   }
 
   private setPlaybackSpeed = (playbackSpeed: number) => {
-    this.mediaRef.current && (this.mediaRef.current.playbackRate = playbackSpeed);
+    this.currentElement && (this.currentElement.playbackRate = playbackSpeed);
+  }
+
+  private get currentElement(): SourceElement | undefined {
+    if (this.videoRef.current) {
```
Comment (zzarcon) [inline src/video.tsx:207]: Shouldn't we use the source type in order to return the current ref?

File: __tests__/index.tsx [hunk 11]
```diff
@@ -1,35 +1,42 @@
...
 
 describe('VideoRenderer', () => {
   const setup = (props?: Partial<VideoProps>) => {
     const src = 'video-url';
-    const children = ((props && props.children) || jest.fn().mockImplementation((video) => {
+    const children = jest.fn().mockImplementation((video) => {
       return video;
-    })) as jest.Mock<any>;
-    const component = mount((
+    }) as jest.Mock<ReturnType<RenderCallback>, Parameters<RenderCallback>>;
```
Comment (zzarcon) [inline __tests__/index.tsx:15]: WOW! So good

File: src/video.tsx (hunk 1 above)
Comment (soswow) [reply src/video.tsx]: I resolved it by having two refs

File: src/video.tsx (hunk 10 above)
Comment (soswow) [reply src/video.tsx]: done

**Pull request merged**